src/main/java/dev/su5ed/sinytra/connector/service/ConnectorLoaderService.java

```diff
@@ -29,16 +31,40 @@ public String name() {
 
     @Override
     public void initialize(IEnvironment environment) {
-        Runnable original = FMLLoader.progressWindowTick;
-        FMLLoader.progressWindowTick = () -> {
-            // Hacky way to run invoke prelaunch outside launch plugin iteration,
-            // prevents crash with REIPC. I'm not going to blame them here since
-            // modlauncher provides users with very limited prelaunch hook accessibility
-            ConnectorEarlyLoader.preLaunch();
+        VarHandle provider = uncheck(() -> TRUSTED_LOOKUP.findStaticVarHandle(ImmediateWindowHandler.class, "provider", ImmediateWindowProvider.class));
+        MethodHandle addReads = uncheck(() -> TRUSTED_LOOKUP.findVirtual(Module.class, "implAddReads", MethodType.methodType(void.class, Module.class, boolean.class)));
+
+        ImmediateWindowProvider original = (ImmediateWindowProvider) provider.get();
+        ImmediateWindowProvider newProvider = new ImmediateWindowProvider() {
+
+            @Override
+            public void updateModuleReads(ModuleLayer layer) {
+                var fm = layer.findModule("forge");
+                if (fm.isPresent()) {
+                    try {
+                        addReads.invoke(original.getClass().getModule(), fm.get(), true);
+                    } catch (Throwable e) {
+                        throw new RuntimeException(e);
+                    }
+                }
+                fm.map(l -> Class.forName(l, "net.minecraftforge.client.loading.NoVizFallback"));
```

Comment:
Is this necessary? It's not being consumed here, and loading a GAME layer class might run mixins early.

Reply:
Not loading a class here leads to mixin plugins loading after `preLaunch`.

Reply:
Let's see. Currently, entrypoints are set up in Connector's mixin plugin, `ConnectorBootstrap`. And so without this line, the mixin plugin would never initialize, which means there would be no preLaunch entrypoint to call. Mixin plugins would only get initialized later at the earliest class load call after preLaunch should've originally fired.

Loading `NoVizFallback`, whose contents are not guaranteed, might also load a minecraft class earlier than expected. I'd rather use a dedicated dummy class to trigger mixin without loading any unwanted classes.

### Re-entrance

After moving `ConnectorEarlyLoader.setup()` from `ConnectorBootstrap` to right before `ConnectorEarlyLoader.preLaunch()`, I ran into another issue regarding duplicate class definition. It occurs during the first initialization of mixin plugins when a plugin loads the class which initially trigerred this load sequence. In our case, before loading andromeda's `Bootstrap` class, it is passed to its mixin plugin which also calls `Bootstrap.getModuleClassPath()`, resulting in the `Bootstrap` class being defined. After finishing running the mixin config, the transforming classloader will try to define `Bootstrap` again without checking if it's already been loaded during the transformation process, resulting in an error.

Fabric fixed this by adding the said check - see https://github.com/FabricMC/fabric-loader/commit/f5b4bf2c1750b1b90934b7a259f1a8fb99b85776. In our case, this fix would have to be applied in SJH, which is not likely to receive updates for 1.20.1. Besides, this would forever remain an issue on Connector's recommended Forge version, 47.1.3.

### The proposal

After playing around for a bit, this is the solution I've come to. Loading `DummyTarget` loads mixin plugins early and eliminates the re-entrance issue, as no mixin plugin references `DummyTarget` itself. Next, we run `setup` and prepare the entrypoints, allowing us to invoke `preLaunch` with its intended behavior. Works great with both Andromeda and REIPC.

What do you think about this approach?

```java
@Override
public void updateModuleReads(ModuleLayer layer) {
    // Invoke mixin on a dummy class to initialize mixin plugins
    // Necessary to avoid duplicate class definition errors when a plugin loads the class that is being transformed
    uncheck(() -> Class.forName("dev.su5ed.sinytra.connector.mod.DummyTarget", false, Thread.currentThread().getContextClassLoader()));
    // Setup entrypoints
    ConnectorEarlyLoader.setup();
    // Run preLaunch
    ConnectorEarlyLoader.preLaunch();
    original.updateModuleReads(layer);
}
```

```java
package dev.su5ed.sinytra.connector.mod;

public class DummyTarget {
}
```

Reply:
I definitely agree with targeting a dummy class instead of Forge ones!

My only concern with this approach is how will this affect hypothetical mods which use language adapters as pseudo entrypoints? Maybe `setup` needs to run before plugins?

Fabric loads adapters before Knot is unlocked and Mixin finishes setup, so Mixin plugins would never load before adapters.

Reply:
Yeah, let's move `setup` above the plugin initializer.